README.md
![cover](https://github.com/RoundedInfinity/fluid_dialog/blob/main/images/cover.png?raw=true)

# Fluid Dialog

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![License: MIT][license_badge]][license_link]

A dialog that can switch between multiple contents and animates size and alignment dynamically.

### Video Demo

![video demo](https://github.com/RoundedInfinity/fluid_dialog/blob/main/images/video.gif?raw=true)

### Features 🎉

- Navigate between different pages
- Animated size and alignment
- Fully customizable

### Usage 🧪

To show a FluidDialog you typically use Flutter's `showDialog` function. You can also use your implementation to show the dialog.

Here's an example:

```dart
showDialog(
 context: context,
 builder: (context) => FluidDialog(
  // Set the first page of the dialog.
  rootPage: FluidDialogPage(
    alignment: Alignment.bottomLeft, //Aligns the dialog to the bottom left.
    builder: (context) => const TestDialog(), // This can be any widget.
  ),
 ),
);
```

To **navigate** in the dialog, use the `DialogNavigator`. It works similarly to the flutter navigator.

This only works for widgets in the FluidDialog

```dart
TextButton(
  onPressed: () => DialogNavigator.of(context).push(
    FluidDialogPage(
      builder: (context) => const SecondDialogPage(),
    ),
  ),
  child: const Text('Go to next page'),
),
```

Take a look at the [example](https://github.com/RoundedInfinity/fluid_dialog/blob/main/example/lib/main.dart) or [demo](https://github.com/RoundedInfinity/fluid_dialog/blob/main/example/lib/demo.dart) for more code.

[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://opensource.org/licenses/MIT
[very_good_analysis_badge]: https://img.shields.io/badge/style-very_good_analysis-B22C89.svg
[very_good_analysis_link]: https://pub.dev/packages/very_good_analysis


CHANGELOG.md
# Changelog

## 2.0.0 - 2023-06-29

* **Chore**: Updated flutter version and dependencies.

## 1.1.1

* **Docs**: Improved docs and readme.
* **Fix**: Updated dependencies

## 1.1.0

* **Breaking**: Renamed `dialogDecoration` to `defaultDecoration`.
* **Feat**: Added the ability to change the decoration of the dialog between routes.
* **Fix**: Fixed transition default animation not working properly.
* **Docs**: Improved docs and readme.
* **Refactor**: Now using [very_good_analysis](https://pub.dev/packages/very_good_analysis).

## 1.0.0

* initial release.


test/fluid_dialog_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:fluid_dialog/fluid_dialog.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/dialog.dart
import 'package:fluid_dialog/src/navigator.dart';
import 'package:flutter/material.dart';

/// {@template fluid_dialog_docs}
/// A dialog that  can switch between multiple contents
/// and animates size and alignment dynamically.
///
/// You can control the dialog with the [DialogNavigator].
/// To use it, get it from the context: `DialogNavigator.of(context)`.
///
/// The content of the dialog is stored in [FluidDialogPage]s.
/// The [rootPage] is the first page in the navigation stack.
/// To change pages and alignment use the [DialogNavigator].
///
/// The size of the dialog is determined by its child elements.
///
/// **Example:**
/// ```dart
/// //Shows a simple dialog.
/// showDialog(
///  context: context,
///  builder: (context) => FluidDialog(
///   // Set the first page of the dialog.
///   rootPage: FluidDialogPage(
///     alignment: Alignment.bottomLeft, //Aligns the dialog to the bottom left.
///     builder: (context) => const TestDialog(), // This can be any widget.
///   ),
///  ),
/// );
///
/// ```
///
/// See also:
/// - [DialogNavigator] for more information about navigating a dialog.
/// - [FluidDialogPage] for creating pages for a dialog.
/// {@endtemplate}
class FluidDialog extends StatelessWidget {
  /// {@macro fluid_dialog_docs}
  const FluidDialog({
    required this.rootPage,
    super.key,
    this.edgePadding = const EdgeInsets.all(12),
    this.alignmentDuration = const Duration(milliseconds: 500),
    this.alignmentCurve = const Cubic(0.2, 0, 0, 1),
    this.sizeDuration = const Duration(milliseconds: 300),
    this.sizeCurve = const Cubic(0.2, 0, 0, 1),
    this.transitionBuilder,
    this.transitionDuration = const Duration(milliseconds: 600),
    this.reverseTransitionDuration = const Duration(milliseconds: 100),
    this.transitionCurve = Curves.easeInOutCubicEmphasized,
    this.reverseTransitionCurve = Curves.easeInOutCubicEmphasized,
    this.defaultDecoration,
  });

  /// The initial page that is shown by the dialog.
  /// This page is shown when the dialog is first created.
  ///
  /// See also:
  /// - [DialogNavigator] for opening pages within a dialog.
  /// - [FluidDialogPage] for creating pages for a dialog.
  final FluidDialogPage rootPage;

  /// The minimum padding on the edges of the screen.
  /// Limits the size of the dialog.
  ///
  /// When this is 0, the dialog can take up the entire screen.
  final EdgeInsetsGeometry edgePadding;

  /// The time it takes for the dialog to change its alignment.
  ///
  /// See also:
  /// - [FluidDialogPage] for setting the alignment of the dialog.
  final Duration alignmentDuration;

  /// The [Curve] used by the alignment animation.
  final Curve alignmentCurve;

  /// The time it takes for the dialog to change its size.
  final Duration sizeDuration;

  /// The [Curve] used by the size change animation.
  final Curve sizeCurve;

  /// The time it takes for the new page to appear.
  final Duration transitionDuration;

  /// The time it takes for the old page to disappear.
  final Duration reverseTransitionDuration;

  /// The [Curve] used for the animation of the appearing page.
  final Curve transitionCurve;

  /// The [Curve] used for the animation of the disappearing page.
  final Curve reverseTransitionCurve;

  /// A builder for the transition used to switch between the different pages.
  ///
  /// This uses a [ZoomPageTransitionsBuilder] by default.
  final AnimatedSwitcherTransitionBuilder? transitionBuilder;

  /// The default style of the dialog.
  ///
  /// Typically a [BoxDecoration].
  final Decoration? defaultDecoration;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: edgePadding,
      child: DialogNavigatorProvider(
        navigator: DialogNavigator(
          context: context,
          pages: ValueNotifier([rootPage]),
        ),
        // Builder used to get new context for DialogNavigatorProvider.
        child: Builder(
          builder: (context) {
            return ValueListenableBuilder(
              valueListenable: DialogNavigator.of(context).pages,
              builder: (context, List<FluidDialogPage> value, child) {
                final page = value.last;

                return AnimatedAlign(
                  duration: alignmentDuration,
                  curve: alignmentCurve,
                  alignment: page.alignment,
                  child: Material(
                    borderOnForeground: false,
                    color: Colors.transparent,
                    child: AnimatedContainer(
                      duration: transitionDuration,
                      curve: transitionCurve,
                      decoration: page.decoration ??
                          defaultDecoration ??
                          BoxDecoration(
                            color: Theme.of(context).colorScheme.surface,
                            borderRadius: BorderRadius.circular(16),
                          ),
                      child: AnimatedSize(
                        duration: sizeDuration,
                        curve: sizeCurve,
                        reverseDuration: sizeDuration,
                        child: AnimatedSwitcher(
                          duration: transitionDuration,
                          reverseDuration: reverseTransitionDuration,
                          switchInCurve: transitionCurve,
                          switchOutCurve: reverseTransitionCurve,
                          transitionBuilder: transitionBuilder ??
                              (child, animation) {
                                // Use default animation
                                return FadeTransition(
                                  opacity: animation,
                                  child: child,
                                );
                              },

                          // Use the current page from the DialogNavigator
                          child: page.builder(context),
                        ),
                      ),
                    ),
                  ),
                );
              },
            );
          },
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:fluid_dialog/fluid_dialog.dart';
import 'package:flutter/material.dart';

import 'demo.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Fluid Dialog Demo',
      theme: ThemeData(
        useMaterial3: true,
        colorSchemeSeed: Colors.redAccent,
      ),
      home: const DemoPage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: OutlinedButton(
          onPressed: () {
            showDialog(
              context: context,
              builder: (context) => FluidDialog(
                rootPage: FluidDialogPage(
                  alignment: Alignment.bottomLeft,
                  builder: (context) => const TestDialog(),
                ),
              ),
            );
          },
          child: const Text('Open Dialog'),
        ),
      ),
    );
  }
}

class TestDialog extends StatelessWidget {
  const TestDialog({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(16.0),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text(
            'Hello there',
            style: Theme.of(context).textTheme.headlineMedium,
          ),
          const Text(
              'This is a dialog. It can transition in from the top, bottom, left, or right. The size is also animated.'),
          TextButton(
            onPressed: () => DialogNavigator.of(context).push(
              FluidDialogPage(
                builder: (context) => const SecondDialogPage(),
              ),
            ),
            child: const Text('Go to next page'),
          ),
        ],
      ),
    );
  }
}

class SecondDialogPage extends StatelessWidget {
  const SecondDialogPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(16.0),
      child: SingleChildScrollView(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(
              'And a bigger dialog',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
            const Text(placeholder),
            TextButton(
              onPressed: () => DialogNavigator.of(context).pop(),
              child: const Text('Go back'),
            ),
            TextButton(
              onPressed: () => DialogNavigator.of(context).close(),
              child: const Text('Close the dialog'),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/navigator.dart
import 'package:flutter/widgets.dart';

/// {@template dialog_navigator_provider_docs}
/// This widget provides a [DialogNavigator] to its children.
///
/// This allows you to call the [DialogNavigator]
/// with `DialogNavigator.of(context)`.
///
/// See also:
/// - [DialogNavigator] for more information about navigating a dialog.
/// - [FluidDialogPage] for creating pages for a dialog.
/// - `FluidDialog` for creating a dialog.
/// {@endtemplate}
class DialogNavigatorProvider extends InheritedWidget {
  /// {@macro dialog_navigator_provider_docs}
  const DialogNavigatorProvider({
    required this.navigator,
    required super.child,
    super.key,
  });

  /// The [DialogNavigator] provided by this widget.
  final DialogNavigator navigator;

  /// The state from the closest instance of this
  /// class that encloses the given context.
  static DialogNavigatorProvider of(BuildContext context) {
    final result =
        context.dependOnInheritedWidgetOfExactType<DialogNavigatorProvider>();
    assert(result != null, 'No DialogNavigatorProvider found in context');
    return result!;
  }

  @override
  bool updateShouldNotify(DialogNavigatorProvider oldWidget) =>
      navigator != oldWidget.navigator;
}

/// {@template dialog_navigator_docs}
/// This class is used to navigate a `FluidDialog`
///
/// This can be used to add or remove pages and set the alignment of the dialog.
///
/// [push] and [pop] are can be used for this.
///
/// {@macro dialog_navigator_example}
/// {@endtemplate}
class DialogNavigator {
  /// {@macro dialog_navigator_docs}
  const DialogNavigator({
    required this.pages,
    required this.context,
  });

  /// The pages stored in the navigation stack.
  ///
  /// This should not be changed manually. Use [push] and [pop] instead.
  final ValueNotifier<List<FluidDialogPage>> pages;

  /// `BuildContext` used for navigation.
  final BuildContext context;

  /// Gets the next [DialogNavigator] from the context.
  ///
  /// This should only be used in a [FluidDialogPage] or
  /// when a [DialogNavigatorProvider] is manually added to the widget tree.
  ///
  /// For info how to use this, see [DialogNavigator].
  static DialogNavigator of(BuildContext context) {
    return DialogNavigatorProvider.of(context).navigator;
  }

  /// Pushes a new [FluidDialogPage] onto the navigation stack.
  ///
  /// Size and alignment will be changed automatically.
  ///
  /// {@template dialog_navigator_example}
  /// **Example:**
  /// ```dart
  /// DialogNavigator.of(context).push(
  ///   FluidDialogPage(
  ///     builder: (context) => Text('Gday'),
  ///  ),
  /// );
  /// ```
  /// {@endtemplate}
  void push(FluidDialogPage page) {
    pages.value = pages.value.toList()..add(page);
  }

  /// Goes back one page.
  ///
  /// Size and alignment will be changed automatically.
  void pop() {
    pages.value = pages.value.toList()..removeLast();
  }

  /// Closes the entire dialog.
  ///
  /// Same as using `Navigator.of(context).pop();`
  void close() {
    Navigator.of(context).pop();
  }

  /// Returns the widget of the last page and wraps it in a [Builder].
  Widget getCurrentWidget() {
    try {
      return Builder(
        builder: (context) => pages.value.last.builder(context),
      );
      // ignore: avoid_catching_errors
    } on StateError {
      throw Exception(
        // ignore: lines_longer_than_80_chars
        'No widget found in pages. Make sure that there is at least one element in the pages list.',
      );
    }
  }
}

/// {@template fluid_dialog_page}
/// A page for the `FluidDialog`.
///
/// This contains the widget shown in the dialog
/// and the alignment of the dialog.
/// {@endtemplate}
class FluidDialogPage {
  /// {@macro fluid_dialog_page}
  const FluidDialogPage({
    required this.builder,
    this.decoration,
    this.alignment = Alignment.center,
  });

  /// A builder for the widget shown in the dialog.
  ///
  /// The `context` contains a [DialogNavigator],
  /// which you can access with `DialogNavigator.of(context)`.
  ///
  /// This can be used to push new pages to the dialog.
  final Widget Function(BuildContext context) builder;

  /// The alignment of the dialog.
  ///
  /// This is `Alignment.center` by default.
  final Alignment alignment;

  /// The [Decoration] used for this page.
  ///
  /// Typically a [BoxDecoration].
  final Decoration? decoration;
}


example/lib/demo.dart
import 'package:animations/animations.dart';
import 'package:fluid_dialog/fluid_dialog.dart';
import 'package:flutter/material.dart';

const String placeholder = '''
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Scripta periculis ei eam, te pro movet reformidans. Eos verear percipit ex, eos ne eligendi inimicus. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Pri viderer tamquam ei. Partiendo adversarium no mea. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. No vis iuvaret appareat. Scripta periculis ei eam, te pro movet reformidans. Scripta periculis ei eam, te pro movet reformidans. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Ad doctus gubergren duo, mel te postea suavitate. Liber nusquam insolens has ei, appetere accusamus intellegam id ius.
''';

class DemoPage extends StatelessWidget {
  const DemoPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Demo'),
        actions: [
          IconButton(
            onPressed: () {
              // The `showModal` function from the animations package is used instead of showDialog
              // because it has a cooler animation.
              showModal(
                context: context,
                builder: (context) => FluidDialog(
                  // Use a custom curve for the alignment transition
                  alignmentCurve: Curves.easeInOutCubicEmphasized,
                  // Setting custom durations for all animations.
                  sizeDuration: const Duration(milliseconds: 300),
                  alignmentDuration: const Duration(milliseconds: 600),
                  transitionDuration: const Duration(milliseconds: 300),
                  reverseTransitionDuration: const Duration(milliseconds: 50),
                  // Here we use another animation from the animations package instead of the default one.
                  transitionBuilder: (child, animation) => FadeScaleTransition(
                    animation: animation,
                    child: child,
                  ),
                  // Configuring how the dialog looks.
                  defaultDecoration: BoxDecoration(
                    color: Theme.of(context).colorScheme.surface,
                    borderRadius: BorderRadius.circular(8.0),
                  ),
                  // Setting the first dialog page.
                  rootPage: FluidDialogPage(
                    alignment: Alignment.topRight,
                    builder: (context) => const InfoDialog(),
                  ),
                ),
              );
            },
            icon: const Icon(Icons.more_vert),
          ),
        ],
      ),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: ListView(
          children: const [
            Placeholder(),
            Text(placeholder),
          ],
        ),
      ),
    );
  }
}

/// A simple example of a dialog
class InfoDialog extends StatelessWidget {
  const InfoDialog({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 200,
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            ListTile(
              title: const Text('Style'),
              leading: const Icon(Icons.color_lens_outlined),
              iconColor: Theme.of(context).colorScheme.onSurface,
            ),
            ListTile(
              title: const Text('Format'),
              leading: const Icon(Icons.text_format_rounded),
              iconColor: Theme.of(context).colorScheme.onSurface,
              // Transform to the next dialog when this is pressed.
              onTap: () => DialogNavigator.of(context).push(
                FluidDialogPage(
                  alignment: Alignment.topRight,
                  builder: (context) => const FormatPage(),
                ),
              ),
            ),
            ListTile(
              title: const Text('Settings'),
              leading: const Icon(Icons.settings_outlined),
              iconColor: Theme.of(context).colorScheme.onSurface,
            ),
            ListTile(
              title: const Text('About'),
              leading: const Icon(Icons.info_outline),
              iconColor: Theme.of(context).colorScheme.onSurface,
              onTap: () => DialogNavigator.of(context).push(
                FluidDialogPage(
                  // This dialog is shown in the center of the screen.
                  alignment: Alignment.center,
                  // Using a custom decoration for this dialog.
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(32),
                    color: Colors.white,
                  ),
                  builder: (context) => const AboutPage(),
                ),
              ),
            ),
            const Divider(),
            const Text('Words: 420'),
            const Text('Reading Time: 6min 9s'),
          ],
        ),
      ),
    );
  }
}

class FormatPage extends StatelessWidget {
  const FormatPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 250,
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            IconButton(
              // Go back to the previous page when the back button is pressed.
              onPressed: () => DialogNavigator.of(context).pop(),
              icon: const Icon(Icons.arrow_back),
              color: Theme.of(context).colorScheme.onSurface,
            ),
            ListTile(
              title: Text(
                'Format',
                style: Theme.of(context).textTheme.titleLarge,
              ),
              subtitle: const Text('Change the format of the page.'),
            ),
            SwitchListTile(
              value: true,
              title: const Text('Enable Something'),
              onChanged: (val) {},
            ),
            SwitchListTile(
              value: false,
              title: const Text('Enable another thing'),
              onChanged: (val) {},
            ),
            const Divider(),
            ListTile(
              title: const Text('Alignment'),
              leading: const Icon(Icons.format_align_left_outlined),
              iconColor: Theme.of(context).colorScheme.onSurface,
            ),
            ListTile(
              title: const Text('Color'),
              leading: const Icon(Icons.format_color_fill_outlined),
              iconColor: Theme.of(context).colorScheme.onSurface,
            ),
          ],
        ),
      ),
    );
  }
}

class AboutPage extends StatelessWidget {
  const AboutPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 800,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                IconButton(
                  onPressed: () => DialogNavigator.of(context).pop(),
                  icon: const Icon(Icons.arrow_back),
                  color: Theme.of(context).colorScheme.onSurface,
                ),
                IconButton(
                  // Close the dialog completely.
                  onPressed: () => DialogNavigator.of(context).close(),
                  icon: const Icon(Icons.close),
                  color: Theme.of(context).colorScheme.onSurface,
                ),
              ],
            ),
            Text(
              'Fluid Dialog Demo',
              style: Theme.of(context).textTheme.titleLarge,
            ),
            Text(
              'Version 1.0.0',
              style: Theme.of(context).textTheme.titleMedium,
            ),
            const Text(placeholder)
          ],
        ),
      ),
    );
  }
}
